design/StableMemory.md

```diff
@@ -144,90 +153,115 @@ If there are no stable variables, we shift the first two words of
-*  each write adjusts max_size and grows memory if necessary
-*  each read adjusts max_size and grows memory if necessary, reading 0 for fresh locations.
+*  aligned with stable-memory, with global word `size` holding logical page count (initially 0 < !size < 2^16).
+*  user are responsible for allocating logical pages.
+*  each load/store does a `size`-related bounds check.
+
+(stable variables aren't maintained in stable memory - they are on the Motoko heap.)
 
-Stable memory layout (between upgrades) :
+Stable memory layout (between upgrades), assuming optional stable variable encoding `v_opt`.
 
 ```
-match stable_vars, mem[0..max-1] with
-  None, mem[0..max-1] ->
-    if ic.stable_size() == 0 then
-    assert max == 0;
-    []
-    else
-    // stable grow mem to page including address | 4 + 4 + max - 1 |
-    [word(0x0)@@max@@mem[4 + 4, ..., max - 1]@@mem[0, ..., 4 + 4 -1 ]
-  Some v, mem[0..max-1] ->
-    let blob = serialise(v) in
-    // stable grow page including address | 4 + |blob| + 4 + max |
-    [size(serialize(v))@@blob@@max@@mem[4 + |blob| + 4, ..., max -1 ]@@mem[0, ..., 4 + |blob| + 4  -1]
+(case !size == 0)
+  (case v == None)
+  <empty> (ie. ic0.stable_size() == 0)
+  (case v_opt = Some v)
+  [N..N+3] StableVariable data len
+  [N+4..(N+4)+len-1] StableVariable data
+  [(N+4)+len-1,..M-1] 0...0 // zero padding
+(case !size > 0)
+[0..3]  0...0
+[4..N-1]  StableMemory bytes
+[N..N+3]  StableVariable data len
+[N+4..(N+4)+len-1] StableVariable data
+[(N+4)+len..M-3] 0...0 // zero padding
+[M-3..M-6] value N/64Ki = !size
+[M-5..M-2] saved StableMemory bytes
+[M-1]  version byte
+
+where N = !size * pagesize // logical memory size
+      M = ic0.stable_size() * pagesize
+      pagesize = 64Kb (2^16 bytes)
+where (len, data) = match v_opt with
+  Some v -> serialize(v,data)
+  None -> (0, [])
 ```
 
 On pre_upgrade
-```
-func pre_upgrade(v_opt, max : word) =
-  match v_opt with
-    null ->
-      if ic.stable_size() == 0 then
-         assert (max == 0);
-         ()
-      else
-      // if necessary, grow mem to size max + 8
-      // save first 2 words to end
-      mem[max, .., max + 7] = mem[0..7];
-      // write marker 0x0 and max to first two words
-      mem.writeWord(0, 0x0)
-      mem.writeWord(4, max)
-  | Some v ->
-    let blob = serialize(v);
-    let size : Word32 = |blob|;
-    assert (size <> 00)
-    //if necessary, grow mem to max + 4 + size + 4
-    // shift first 4 + size + 4 words to end
-    mem[max, max + 4 + size + 4 - 1] = mem[0.. 4 + size + 4  -1];
-    mem.writeWord(0, size)       // save non-zero size of blob
-    mem.writeBlob(4, blob)       // save blob
-    mem.writeWord(4 + size, max) // save max
+
+```ocaml
+func save v_opt : value option =
+  let len, data = match v_opt with
+    | Some v -> serialize(v)
+    | None -> 0, []
+  in
+  if !size == 0 then
+    assert (ic0.stable_size() == 0);
```

Comment:
Is that correct? It might already have been grown in previous upgrades.

Reply:
Quite right, removing the assert.